src/Storages/StorageMergeTree.cpp

```diff
@@ -1831,9 +1855,16 @@ MutationCommands StorageMergeTree::getFirstAlterMutationCommandsForPart(const Da
     std::lock_guard lock(currently_processing_in_background_mutex);
 
     auto it = current_mutations_by_version.upper_bound(part->info.getDataVersion());
-    if (it == current_mutations_by_version.end())
-        return {};
-    return it->second.commands;
+    while (it != current_mutations_by_version.end())
```

Comment:
Not sure if it's the problem, but it seem like in the pessimistic scenario (100000 parts of a version lower than max), that `while` will be executed quite often...

Reply:
Fixed in 47c9fd519c0604371f3b18577236151df18aae3d .